node_test.go

```diff
@@ -1312,3 +1315,130 @@ func TestNode_DeleteMultiple_Query(t *testing.T) {
 	}), ".", []interface{}{}))
 	t.Run("RootNodeUnknown", deleteMultipleTest(dasel.New(false), ".", map[string]interface{}{}))
 }
+
+// TestNode_NewFromFile tests parsing from file
+//
+// Not sure what to test here:
+// should all file/parser tests be duplicated?
+// E.g.: TestLoadFromFile, TestNewReadParserFromString
+func TestNode_NewFromFile(t *testing.T) {
+	tests := []struct {
+		name   string
+		file   string
+		parser string
+		err    error
+	}{
+		{
+			name:   "File exists and valid parser specified",
+			file:   "./tests/assets/example.json",
+			parser: "json",
+			err:    nil,
+		},
+		{
+			name:   "File exists and invalid parser specified",
+			file:   "./tests/assets/example.json",
+			parser: "bad",
+			err:    storage.UnknownParserErr{Parser: "bad"},
+		},
+		{
+			name:   "File doesn't exist and valid parser specified",
+			file:   "no_such_file",
+			parser: "json",
+			err:    errors.New("could not open file: open no_such_file"),
+		},
+		{
+			name:   "File doesn't exist and invalid parser specified",
+			file:   "no_such_file",
+			parser: "bad",
+			err:    storage.UnknownParserErr{Parser: "bad"},
+		},
+	}
+
+	for _, testCase := range tests {
+		tc := testCase
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := dasel.NewFromFile(tc.file, tc.parser)
+
+			if tc.err == nil && err != nil {
+				t.Errorf("expected err %v, got %v", tc.err, err)
+				return
+			}
+			if tc.err != nil && err == nil {
+				t.Errorf("expected err %v, got %v", tc.err, err)
+				return
+			}
+			if tc.err != nil && err != nil && !strings.Contains(err.Error(), tc.err.Error()) {
+				t.Errorf("expected err %v, got %v", tc.err, err)
+				return
+			}
+		})
+	}
+}
+
+// TestNode_Write tests writing to Writer
+//
+// Not sure what to test here:
+// should all file/parser tests be duplicated?
+// E.g.: TestWrite, TestNewWriteParserFromString
```

Comment:
Same as before

Reply:
I've extended tests to cover all possible combinations, but some `XML` tests are failing. I'm not sure how (or if) this should be addressed in the scope of this PR. The `xml` writer doesn't do compact/escape.

```none
--- FAIL: TestNode_Write (0.00s)
    --- FAIL: TestNode_Write/XML,_pretty,_escape (0.00s)
        node_test.go:1760: Node.Write() = <key>
              <value><&></value>
            </key>
             want <key>
              <value>&lt;&amp;&gt;</value>
            </key>
    --- FAIL: TestNode_Write/XML,_compact,_escape (0.00s)
        node_test.go:1760: Node.Write() = <key>
              <value><&></value>
            </key>
             want <key><value>&lt;&amp;&gt;</value></key>
    --- FAIL: TestNode_Write/XML,_compact,_no_escape (0.00s)
        node_test.go:1760: Node.Write() = <key>
              <value><&></value>
            </key>
             want <key><value><&></value></key>
    --- FAIL: TestNode_Write/Pre-escaped_XML,_pretty,_escape (0.00s)
        node_test.go:1760: Node.Write() = <key>
              <value>&lt;&amp;&gt;</value>
            </key>
             want <key>
              <value>&amp;lt;&amp;amp;&amp;gt;</value>
            </key>
    --- FAIL: TestNode_Write/Pre-escaped_XML,_compact,_escape (0.00s)
        node_test.go:1760: Node.Write() = <key>
              <value>&lt;&amp;&gt;</value>
            </key>
             want <key><value>&amp;lt;&amp;amp;&amp;gt;</value></key>
    --- FAIL: TestNode_Write/Pre-escaped_XML,_compact,_no_escape (0.00s)
        node_test.go:1760: Node.Write() = <key>
              <value>&lt;&amp;&gt;</value>
            </key>
             want <key><value>&lt;&amp;&gt;</value></key>
FAIL
exit status 1
FAIL    github.com/tomwright/dasel      0.034s
```

Reply:
Thanks for that. That's fine in this use-case since the parser itself doesn't support it 👍